src/if_statements.py

```diff
@@ -617,6 +691,22 @@ def build_flowgraph_between(context: Context, start: Node, end: Node) -> Body:
     while curr_start != end:
         assert not isinstance(curr_start, TerminalNode)
 
+        if not skip_loop_detection and isinstance(curr_start, ConditionalNode):
+            # breakpoint()
+            do_while_loop = detect_loop(context, curr_start, end)
+            if do_while_loop:
+                # breakpoint()
+                body.add_do_while_loop(do_while_loop)
+                # Move past the loop:
+                if curr_start.loop and curr_start.loop.is_self_loop():
+                    curr_start = curr_start.fallthrough_edge
+                else:
```

Comment:
It's possible that the conditional at the head of the loop is unrelated to the loop itself, unfortunately:

```diff
diff --git a/tests/end_to_end/loop_with_if/orig.c b/tests/end_to_end/loop_with_if/orig.c
index a42150f..172b179 100644
--- a/tests/end_to_end/loop_with_if/orig.c
+++ b/tests/end_to_end/loop_with_if/orig.c
@@ -9,5 +9,13 @@ int test(int length) {
         }
     }
 
+    do {
+        if (i == 7) {
+            i *= 3;
+        } else {
+            i += 6;
+        }
+    } while (i < length * 2);
+
     return i;
 }
```

![2021-06-24-021128_628x447_scrot](https://user-images.githubusercontent.com/113075/123211532-94662280-d491-11eb-8f92-658c0eac1525.png)

Here, Node 7 is the head of the loop, and a conditional, but the conditional is the if statement and is unrelated to the loop itself. (The loop conditional is in node 11)